### ISSUE TITLE
problem with barcode matching

### ISSUE THREAD

**Issue by scZfish** [landau-lab/Gotcha#10]
I am having a problem with barcode matching when using the MutationCalling function:

MutationCalling(out = "/home/nathan.lawson-umw/scATACruns/2023_9_28_pilotATACgeno1/gotchaPilot/fastqOutSubset/Split/Filtered/aa/",
+                      barcodes.file.path = "/home/nathan.lawson-umw/scATACruns/2023_9_28_pilotATACgeno1/scATACgeno1/outs/filtered_peak_bc_matrix/barcodes.tsv",
+                      wt.max.mismatch = 0,
+                      mut.max.mismatch = 0,
+                      keep.raw.reads = F,
+                      reverse.complement = T,
+                      which.read = "R1",
+                      testing = F,
+                      primer.sequence = "CCTGCCTACGCAACAAGATGTGCTT",
+                      primed.max.mismatch = 3,
+                      wt.sequence = "CGA",
+                      mut.sequence= "CAA",
+                      mutation.start = 31,
+                      mutation.end = 34
+ )
------- BEGIN MUTATION CALLING FUNCTION -------
------- FASTQ FILES LOADED -------
------- SEQUENCES OBTAINED FROM FASTQ FILES -------
Data merged for each lane...
------- DATAFRAME FOR EACH SAMPLE GENERATED -------
------- PRIMED READS IDENTIFIED -------
number of starting reads = 24177
number of primed reads = 22574
% of primed reads = 93.37
------- BARCODES IDENTIFIED -------
------- WARNING: Cell barcodes (CB) have been reverse complemented due to sequencer used -------
------- BARCODE MATCHING BEGINNING -------
------- BARCODE MATCHING DONE -------
total number of primed barcodes = 22574
total number of primed matched barcodes = 0
% barcode matching = 0
Error in `$<-.data.frame`(`*tmp*`, "ReadGenotype", value = NA) : 
  replacement has 1 row, data has 0

This is using a subset of 25K reads out of the total to check if the package is working for me. Most of my reads have the primer and I can see both mutant and wild type sequences if I just browse through the fastq file. Also, if I pull out the barcodes from the R2 fastq.gz, reverse complement them and use the list as a text file to grep the barcodes.tsv file, it looks like about 260 are matching. So, there are definitely matching barcodes in my 25K subset.  I tried using a reverse complemented R2 fastq.gz, as well as using the reverse complement for all reads as primary input (separate tries), but get the same error.

Session info:

> sessionInfo()
R version 4.2.2 Patched (2022-11-10 r83330)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Gotcha_1.0.1

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.11                 stringdist_0.9.10           lattice_0.20-45             deldir_1.0-9               
 [5] png_0.1-8                   Rsamtools_2.14.0            Biostrings_2.66.0           utf8_1.2.3                 
 [9] rslurm_0.6.2                R6_2.5.1                    GenomeInfoDb_1.34.9         ShortRead_1.56.1           
[13] stats4_4.2.2                pillar_1.9.0                zlibbioc_1.44.0             rlang_1.1.1                
[17] rstudioapi_0.14             minqa_1.2.6                 nloptr_2.0.3                S4Vectors_0.36.2           
[21] Matrix_1.6-1.1              reticulate_1.32.0           splines_4.2.2               lme4_1.1-34                
[25] BiocParallel_1.32.6         readr_2.1.4                 RCurl_1.98-1.12             DelayedArray_0.24.0        
[29] compiler_4.2.2              rtracklayer_1.58.0          pkgconfig_2.0.3             BiocGenerics_0.44.0        
[33] tidyselect_1.2.0            SummarizedExperiment_1.28.0 tibble_3.2.1                GenomeInfoDbData_1.2.9     
[37] IRanges_2.32.0              codetools_0.2-19            matrixStats_1.0.0           XML_3.99-0.14              
[41] fansi_1.0.4                 crayon_1.5.2                dplyr_1.1.3                 tzdb_0.4.0                 
[45] GenomicAlignments_1.34.1    MASS_7.3-58.2               bitops_1.0-7                grid_4.2.2                 
[49] nlme_3.1-162                jsonlite_1.8.7              lifecycle_1.0.3             magrittr_2.0.3             
[53] cli_3.6.1                   XVector_0.38.0              hwriter_1.3.2.1             latticeExtra_0.6-30        
[57] generics_0.1.3              vctrs_0.6.3                 boot_1.3-28                 rjson_0.2.21               
[61] restfulr_0.0.15             RColorBrewer_1.1-3          tools_4.2.2                 interp_1.1-4               
[65] Biobase_2.58.0              glue_1.6.2                  hms_1.1.3                   MatrixGenerics_1.10.0      
[69] jpeg_0.1-10                 yaml_2.3.7                  parallel_4.2.2              GenomicRanges_1.50.2       
[73] BiocIO_1.8.0

**Comment by Theob0t** [comment]
Hi, 
Please try using the `/outs/singlecell.csv` file from the cellranger ATAC output folder as your `barcodes.file.path` input and see if that fixes the error.
Thanks,

**Comment by scZfish** [comment]
Thanks. That worked, but since that file is much larger, it ends of taking much more time.

The error I got might be because the barcodes.tsv and singlecell.csv files look different? The .csv file has a header along with count data, while the .tsv file only has the barcode.  

I used a join between the outs/singlecell.csv file and filtered_peak_bc_matrix/barcodes.tsv to generate a more restricted singlecell.csv file and then made sure I had the header. This ended up working and much faster than using the full singlecell.csv. If I remove the header, then it threw an error. 

In any case, it is working now! Thanks.
